### DISCUSSION TITLE
Apache Passenger config ?

### DISCUSSION THREAD

**Discussion by rubydesign** [mastodon/mastodon#27648] [Q&A]
Hi all,
so i am trying to dip into the federated world. I already have a dedicated server and 5 of my own apps running on it with apache + passenger. Since i have done that combo for years i do not know docker/nginx, but would have no problem getting the basic app to run.

The reason i am asking is just that i looked at the nginx config and it has all the cache and streaming stuff in there which i am not so certain about. I think i might get it done, but it just seems worth the question if someone does not have a ready config. I mean my situation of already having a server with apache/passenger can not be _that_ unique, given there are supposed to be 10k servers.

PS. my user is not called mastodon, can anyone say how important that is?

Anyway, any help appreciated, if not i will post my results later

**Comment by vmstan** [comment]
The user does not have to be Mastodon.

**Comment by ViniciusFSSilva** [comment]
> The user does not have to be Mastodon.

**Comment by ViniciusFSSilva** [comment]
Yes

**Comment by rubydesign** [comment]
Ok,  i have my instance up. The apache was not so hard, but the services and deployment took a while. Also in reply to previous user question it should be mentioned that service examples DO have the user mastodon hardcoded. But more on services below.

Apache conf, using passenger. Comments about extra cache control welcome.
```
<VirtualHost *:443>
    ServerName feenix.community
    # Tell Apache and Passenger where your app's 'public' directory is
    # I use mina, but capistrano is the same in that it has a current dir for the current release
    DocumentRoot /home/feenix/mastodon/current/public

    # I use passenger, this block must be done with ProxyPass when using puma
    PassengerRuby /home/feenix/.rbenv/versions/3.2.2/bin/ruby
    # Relax Apache security settings
    <Directory /home/feenix/mastodon/current/public>
      Allow from all
      Options -MultiViews
      # Uncomment this if you're on Apache ≥ 2.4:
      Require all granted
    </Directory>

    # Node streaming, see service section below
    ProxyPass /api/v1/streaming/ ws://localhost:4000/
    ProxyPassReverse /api/v1/streaming/ ws://localhost:4000/

</VirtualHost>
```
That is what i start with. Then i do the certbot command, which inserts rewrites into the 80 config, and creates an ssl 443 config with the same. I usually delete the passenger/app stuff from the 80 port and just leave the rewrite/redirect stuff in.

For the services there were examples that required sudo. I (and debian) disagree with sudo in that if a user is compromised it should not have sudo right. So i changed the service files, see below. 
With those files sudo only needs to be invoked once for `loginctl enable-linger USERNAME`  , all other commands are by the user.
For sidekiq:
`systemctl --user enable mastodon-sidekiq.service`
`systemctl --user start mastodon-sidekiq.service`
For streaming:
`systemctl --user enable mastodon-streaming.service`
`systemctl --user start mastodon-streaming.service`

For that to work, below files have to be in `~/.config/systemd/user` . Notice especially the wanted-by , as it is now a user service.
I will post a deployment file seperately as this is getting long, with all this deployment can be done by the user without sudo!!

mastodon-sidekiq.service (from sidkiq github):

```
# This file tells systemd how to run Sidekiq as a 24/7 long-running daemon.
#
# Customize this file based on your bundler location, app directory, etc.
#
# If you are going to run this as a user service (or you are going to use capistrano-sidekiq)
# Customize and copy this to ~/.config/systemd/user
# Then run:
#   - systemctl --user enable sidekiq
#   - systemctl --user {start,stop,restart} sidekiq
# Also you might want to run:
#   - loginctl enable-linger username
# So that the service is not killed when the user logs out.
#
# If you are going to run this as a system service
# Customize and copy this into /usr/lib/systemd/system (CentOS) or /lib/systemd/system (Ubuntu).
# Then run:
#   - systemctl enable sidekiq
#   - systemctl {start,stop,restart} sidekiq
#
# This file corresponds to a single Sidekiq process.  Add multiple copies
# to run multiple processes (sidekiq-1, sidekiq-2, etc).
#
# Use `journalctl -u sidekiq -rn 100` to view the last 100 lines of log output.
#
[Unit]
Description=mastodon-sidekiq
# start us only once the network and logging subsystems are available,
# consider adding redis-server.service if Redis is local and systemd-managed.
After=syslog.target network.target

# See these pages for lots of options:
#
#   https://www.freedesktop.org/software/systemd/man/systemd.service.html
#   https://www.freedesktop.org/software/systemd/man/systemd.exec.html
#
# THOSE PAGES ARE CRITICAL FOR ANY LINUX DEVOPS WORK; read them multiple
# times! systemd is a critical tool for all developers to know and understand.
#
[Service]
#
#      !!!!  !!!!  !!!!
#
# As of v6.0.6, Sidekiq automatically supports systemd's `Type=notify` and watchdog service
# monitoring. If you are using an earlier version of Sidekiq, change this to `Type=simple`
# and remove the `WatchdogSec` line.
#
#      !!!!  !!!!  !!!!
#
Type=notify
# If your Sidekiq process locks up, systemd's watchdog will restart it within seconds.
WatchdogSec=10

WorkingDirectory=/home/feenix/mastodon/current
# If you use rbenv:
ExecStart=/bin/bash -lc 'exec /home/feenix/.rbenv/shims/bundle exec sidekiq -e production'

# Use `systemctl kill -s TSTP sidekiq` to quiet the Sidekiq process
# Greatly reduce Ruby memory fragmentation and heap usage
# https://www.mikeperham.com/2018/04/25/taming-rails-memory-bloat/
Environment=MALLOC_ARENA_MAX=2

# if we crash, restart
RestartSec=2
Restart=always

# This will default to "bundler" if we don't specify it
SyslogIdentifier=sidekiq

[Install]
WantedBy=default.target 

```
 mastodon-streaming.service , only single instance on port 4000 :

```
[Unit]
Description=mastodon-streaming on port 4000
After=network.target
# handles using `systemctl restart mastodon-streaming`

[Service]
Type=simple

WorkingDirectory=/home/feenix/mastodon/current
Environment="NODE_ENV=production"
Environment="PORT=4000"
ExecStart=/usr/bin/node ./streaming
TimeoutSec=15
Restart=always
LimitNOFILE=65536

[Install]
WantedBy=default.target 

```
